w3lib/url.py

```diff
@@ -19,6 +19,28 @@
 from w3lib.util import to_bytes, to_native_str, to_unicode
 
 
+def _encode_netloc(onetloc):
+    """
+    :type onetloc: unicode
+    :rtype: unicode
+    """
+    try:
+        idx = onetloc.rfind(u':')
```

Comment:
Thanks for the tests.
My https://github.com/scrapy/w3lib/pull/98#discussion_r142931555 still holds though.
Have you considered using `SplitResult` `.hostname` and `.port` attributes?

Reply:
https://docs.python.org/2/library/urlparse.html#urlparse-result-object. no `hostname` or `port`. just `netloc`

Reply:
Current behavior:

```py
>>> w3lib.url.safe_url_string('https://www.baIdu.com')
'https://www.baIdu.com'  # I
>>> w3lib.url.canonicalize_url('http://www.baidu.com:80000000')
'http://www.baidu.com:80000000/'
```

when `_encode_netloc()` using `SplitResult` `.hostname` and `.port` attributes:

```py
>>> w3lib.url.safe_url_string('https://www.baIdu.com')
'https://www.baidu.com'
>>> w3lib.url.canonicalize_url('http://www.baidu.com:80000000')
'http://www.baidu.com/'
```

because [python 3.5](https://docs.python.org/3.5/library/urllib.parse.html#urllib.parse.urlsplit) says:

> `hostname` is Host name (**lower case**)...

Is this what we want?

Reply:
I'm fine with lowercasing the hostname. But it is indeed a change in behavior.

Reply:
```
>>> w3lib.url.canonicalize_url('http://www.baidu.com:80000000')
'http://www.baidu.com/'
```
what's the reason for the "lost" port number?

Reply:
[python 3.5](https://docs.python.org/3.5/library/urllib.parse.html#urllib.parse.urlsplit) says: port is None when value if not valid. and the new  `_encode_netloc()`:

```py
def _encode_netloc(parts):
	try:
		hostname = to_unicode(parts.hostname.encode('idna'))
		netloc = hostname if parts.port is None else '%s:%s' % (hostname, parts.port)
	except UnicodeError:
		netloc = parts.netloc
	return netloc
```

Reply:
It's a shame that `urlsplit` and `urlparse` in Python 2.7 and 3.5 simply swallow an invalid port number.
Python 3.6 does report an error:
```
$ python
Python 3.6.2 (default, Aug 24 2017, 10:48:24) 
[GCC 6.3.0 20170406] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from urllib.parse import urlsplit
>>> t = urlsplit('http://www.BaidU.com:654444')
>>> t.port
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/urllib/parse.py", line 169, in port
    raise ValueError("Port out of range 0-65535")
ValueError: Port out of range 0-65535
```